### DISCUSSION TITLE
how to navigate to a given page number

### DISCUSSION THREAD

**Discussion by abubelinha** [py-pdf/fpdf2#987] [Q&A] [answered]
I suppose most of the time fpdf is used in a way that pages are created in order.
So you `add_page()` and then you add content to that page, and then you create the next page and so on ...

But what if I need to add something to a previous page?
For example, creating a table of contents or index in one of the initial pages, listing chapters and their page numbers.
Maybe you don't know those page numbers in advance because chapter lengths may depend on several factors (dynamic content coming from a database, or even selected font type and size).

In that context, it would be easier to collect that info (chapter titles and page numbers) during document creation, and finally navigate back to your TOC page (i.e., page 2) and write the list of chapters at the end of your script.

Is this somehow possible?

Thanks
@abubelinha

**Comment by andersonhc** [comment]
If you mark the chapters using outlines (bookmarks) you can insert a placeholder using insert_toc_placeholder().
Documentation is here: https://py-pdf.github.io/fpdf2/DocumentOutlineAndTableOfContents.html

**Reply by abubelinha** [reply]
Thanks a lot for the info. 
I didn't know about that possibility and I will certainly use it at some point.

That solves the TOC example I mentioned.  But it doesn't answer the title question. 
Is it possible to go back an add stuff to a previously added page?

Actually my idea was not exactly a TOC (I mentioned that as an example of needing to go back, because it was a common use case).

I am thinking about drawing or including images in different pages of a document, but the initial image content (1st page) will summarize data used when drawing other graphs in the next pages. 
It would be much easier for me to summarize data as I am processing it and drawing each part. 
And in the end, when I have the sums, averages, means, whatever ... go back to the initial page and draw my "summary-graph" there.

So I am not dealing with text, but graphical content. 
Not sure if this can also be solved with `insert_toc_placeholder()` (still have to read it).
But anyway I'd like to know if it is somehow possible to go back and edit a previous page.

**Reply by gmischler** [reply, marked as answer]
In its core, fpdf2 is still a very simple system. It was designed a long time ago to process pages in a linear fashion, and that aspect hasn't significantly changed since then. You're looking at a straightforward PDF generator, and not at a content management system.

There are some ways to "hack" around the limitations of the system, either through placeholders like for TOCs or the number of pages, or by manipulating the internal counter `self.page`. The latter is probably what you're looking for here, but you need to be aware that it is really more of a workaround than a real feature, and may result in unintended consequences if not handled with care. Those consequences may also change in the future without notice, since that attribute is not publicly documented.

If you have any chance, then I'd recommend to preprocess your data, and figure out the sequence and the details of the content to render in advance. It may seem like a burden at first, but will really make your life a lot easier in the long run.

**Reply by abubelinha** [reply]
>  It was designed a long time ago to process pages in a linear fashion

Thanks for confirming.
 
Then I agree: preprocessing, storing somewhere and then rendering is the way to go.
That's what I suspected but I wanted to be sure, just in case I could avoid storage.
